### PR TITLE
[testing][wip]

### PR DESCRIPTION
**Pull request by zou3519**
Another vmap prototype.

cc @cpuhrsch

### CODE REVIEW

File: aten/src/ATen/Batching.cpp [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+  auto actual_dim = actualDim(dim, maybe_batch_dim);
+  optional<int64_t> new_batch_dim = nullopt;
+  if (maybe_batch_dim) {
+    auto bdim = *maybe_batch_dim;
+    new_batch_dim = bdim < actual_dim ? bdim : bdim + 1;
+  }
+  return { tensor.unsqueeze(actual_dim), new_batch_dim };
+}
+
+std::pair<Tensor,optional<int64_t>> conv2d_batching_rule(
```
Comment (zou3519) [inline aten/src/ATen/Batching.cpp]: conv2d_batching_rule

File: aten/src/ATen/Batching.cpp [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+  }
+  if (bias_bdim) {
+    TORCH_CHECK(false, "NYI: conv2d_batching_rule for batched bias");
+  }
+  auto result_dim = minDim(self, self_bdim);
+  auto self_ = moveBatchDimToFront(self, self_bdim, result_dim);
+  auto self_sizes = self_.sizes(); 
+
+  auto self_4d = self_.flatten(0, 1);
+  auto result_4d = at::conv2d(
```
Comment (zou3519) [inline aten/src/ATen/Batching.cpp]: conv2d kernel call

File: torch/batching.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import torch
+import functools
+from torch import Tensor
+
+VMAP_LEVEL = 0
```
Comment (cpuhrsch) [inline torch/batching.py]: Why does this need to be a global?

Comment (zou3519) [reply torch/batching.py]: Alternatively, it could be a thread_local. This value is keeping track of how many nested vmaps we are inside.

**Pull request closed without merging**